### ISSUE TITLE
Features not behaving additively

### ISSUE THREAD

**Issue by lukashermansson** [withoutboats/heck#42]
I ran into issues using a library that had this library at two places in the dependency graph.

as noted for the example below when the Unicode feature is enabled it changes the observed behavior of seemly non-Unicode strings

I would expect the bellow test to pass and have the same behavior both with:
cargo test
cargo test --features unicode (outputs Entity.allay.ambientWithItem instead of the behavior bellow)
```
t!(test9: "entity.allay.ambient_with_item" => "EntityAllayAmbientWithItem");
```

**Comment by jplatte** [comment]
That's interesting, I thought punctuation was always stripped with older versions of heck, and I don't think anything changed w.r.t. when the unicode-aware impl (previously the only one available) is used. I don't really have time to maintain this crate anymore, but a PR changing the unicode impl to behave as documented (dropping punctuation) would be appreciated.

**Comment by lukashermansson** [comment]
I looked into this a bit further, the "unicode" feature is probably the feature that has the correct behavior (if the source of truth is http://www.unicode.org/reports/tr29/ and assuming that "unicode_segmentation" has implemented the spec correctly). 

as noted in the screenshot. a period "." between the numbers in the example bellow from the "unicode_segmentation" crate is not treated as a word separator, and I think it's the same deal with the code mentioned in the issue.
![image](https://user-images.githubusercontent.com/7715996/234900616-410036a6-3c72-4c76-9c6b-0ab5d36e50e9.png)

I think overall its this hints towards a much larger issue that behavior will continue to differ if there is any difference in this crates implementation and the external dependencies logic for dealing with word separation
I can think of a couple ways to solve this issue:
- let this crate drop support for the non "unicode" feature, because  if they ever differ in behavior that is a big problem
- provide two different apis one with unicode behaviour and one with own ("best effort") implementation to detect word separation. importantly, it should be two different apis to not break in cases like above.

I don't think just patching the punctuation in this case would be the last issue related to this (this crate would then have to behave the same way as the unicode_segmentation crate behaves when it comes to word detection, I have no idea if this is something that is achievable).

**Comment by jplatte** [comment]
The readme specifically calls out that this crate drops punctuation though. I don't think what the unicode standard says is super relevant for this.

**Comment by lukashermansson** [comment]
Alright! That seems fine, I don't think I will be able to implement a fix myself. So if you or somebody else want's to implement the described fix, this is an open issue.

**Comment by rj00a** [comment]
> I thought punctuation was always stripped with older versions of heck

This is not the case. c501fc95db91ce20eaef248a511caec7142208b4 introduced the optional unicode support, and prior to this the `unicode_words` function was used directly. The current non-unicode behavior is an oversight by the author of that commit.

> The readme specifically calls out that this crate drops punctuation though

It dropped punctuation in some situations, but not always.

```rust
fn check(a: &str, b: &[&str]) {
    assert_eq!(&a.unicode_words().collect::<Vec<_>>(), b);
}

check("foo.bar", &["foo.bar"]);
check("foo..bar", &["foo", "bar"]);
check(".foobar.", &["foobar"]);
```

So I think there is some wiggle room in the interpretation of the readme.

Regardless, having the semantics of the case conversion functions change with the addition of a feature flag seems like a major footgun to me, and is possibly a misuse of Rust's feature flags. It reminds me of the problems people have in C where the behavior of `printf` can change in different locales, causing programs to break.

I think the best course of action here is to revert c501fc95db91ce20eaef248a511caec7142208b4 completely like @lukashermansson mentioned and release a new major version. If @jplatte is OK with this plan, I would be happy to submit a PR.

**Comment by rj00a** [comment]
After thinking about the problem for a bit, I agree that the new behavior should be similar to the current non-unicode impl.

**Comment by Bromeon** [comment]
First, thanks for all the effort that went into this great library. It's really to-the-point, has useful conventions and I love the zero dependencies.

I stumbled upon this issue as well (although interestingly, it helped find me another bug 🙂).

Despite being fixed since over half a year, there is unfortunately no stable crates.io release since. 
Is there a `v0.5.0` or so planned sometime over the next months?
Or what's the risk of upgrading to `v0.5.0-rc.1`?

**Comment by jplatte** [comment]
Maintenance of this crate is not a big priority to me, but I just released v0.5.0. I have not been getting any feedback on the rc release so I assume it's good.
